dags/synapse-dataset-to-croissant.py

```diff
@@ -165,17 +199,559 @@ def set_up_logging() -> Logger:
+                "contentUrl": f"https://www.synapse.org/Synapse:{file.id}.{file.version_number}",
+                "encodingFormat": "application/json",
+                "md5": file_md5,
+                "sha256": "unknown",
+            }
+        )
+    return distribution_files
+
+
+def construct_record_set_section_for_files(files_attached_to_dataset: List[File]) -> Dict[str, str]:
+    """
+    Construct the record set section for the files attached to the dataset. This is used
+    to extract the keys from the annotations of the files and create a Field for each
+    key.
+
+    Arguments:
+        files_attached_to_dataset: The list of files attached to the dataset.
+
+    Returns:
+        The record set section for the files attached to the dataset.
+    """
+    unique_annotation_keys = set()
+
+    for file in files_attached_to_dataset:
+        file: File = file
+        file_annotations = file.annotations.keys()
+        for key in file_annotations:
+            unique_annotation_keys.add(key)
+
+    metadata_fields = []
+    # Sort the set of unique_annotation_keys ignoring case, but keep the case in the result
+    unique_annotation_keys = sorted(unique_annotation_keys, key=str.casefold)
+
+    for key in unique_annotation_keys:
+        metadata_fields.append(
+            {
+                "@type": "Field",
+                "@id": f"metadata/{key}",
+                "name": key,
+                "description": "",
+                "dataType": "sc:Text",
+                "source": {
+                    "fileObject": {"@id": "metadata"},
+                    "extract": {"column": key}
+                }
+            }
+        )
+
+    return {
+        "@type": "RecordSet",
+        "@id": "default",
+        "name": "default",
+        "description": "Metadata for the dataset",
+        "field": metadata_fields
+    }
+
+
+def extract_s3_objects_to_delete(bucket_objects: List[str], dataset_collections_to_consider_for_deletion: List[str], combined_dataset_collection_and_datasets: List[Dict[str, str]]) -> List[str]:
+    """
+    Extract the S3 objects to delete from the S3 bucket. This is done by retrieving
+    all of the objects in the S3 bucket and checking if the object matches one of the
+    datasets that will be kept after the full DAG as been run. Every object in the S3
+    bucket that matches the pattern `_syn####.###_` and
+    `_datasetCollection_syn####_` will be considered for deletion. The DAG param
+    `dataset_collections_for_cleanup` is used to filter the objects that will be
+    considered for deletion as the second filter.
+
+    The last filter enforces that each object in the S3 bucket must match a dataset
+    id, dataset version, dataset collection, and dataset name in the
+    `combined_dataset_collection_and_datasets` list. If the object does not match
+    any of the objects in the list, then it will be marked for deletion.
+
+    Arguments:
```

Comment:
here this is also missing a docstring for parameter `combined_dataset_collection_and_datasets`?